knowledge-base/adr/0031-resilience-options.md

```diff
@@ -61,310 +61,153 @@ So we have something for circuit breaker and timeout.
-- For circuit breaker, we will use [opossum](https://www.npmjs.com/package/opossum)
-
-This determines the possible options you can set.
-The `RetryOptions` and `CircuitBreakerOptions` could be used to overwrite the default values.
-We will introduce a context passed to the resilience implementation to give information on the call:
+The implementation is based on a middleware approach:
 
 ```ts
-type RequestContext = {
-    category: 'xsuaa' | 'destination' | 'target',
-    url?: string,
-    headers?: Record<string,string>,
-    jwt?: string,
-    destination?: Destination,
-    method: 'GET' | 'POST' | 'DELETE',
-    ...
-}
-```
-
-The same object is also used to determine and adjust the options
-
-```ts
-type RetryOptions = false | AsyncRetryLibOptions;
-type CircuitBreakerOptions = false | Omit<OpossumLibOptions, 'timeout'>; // timeout is not handled by opossum
-type TimeoutOptions = false | number;
-
-interface OpossumLibOptions {
-  timeout?: StringValue | false | undefined; // default 10sec
-  errorThresholdPercentage?: number | undefined; // default 50
-  volumeThreshold?: number | undefined; // default 10
-  resetTimeout?: number | undefined; // default 30000
-  isolationStragtegy?: IsolationStrategy; // default tenant
-}
-
-interface AsyncRetryLibOptions {
-  retries?: number; // default 10
-  factor?: number; // default  2.
-  minTimeout?: StringValue; // default 1000 ms. See https://github.com/vercel/ms
-  maxTimeout?: StringValue; // default Infinity. See https://github.com/vercel/ms
-  randomize?: boolean; // default true.
-  onRetry: (e: Error) => {}; // default undefined
-}
-
-// in the resilience call:
-type ResilienceMiddlewareOptions = {
-  timeout: (context?: RequestContext) => TimeoutOptions;
-  retry: (context?: RequestContext) => RetryOptions;
-  circuitBreaker: (context?: RequestContext) => CircuitBreakerOptions;
+type MiddlewareInOut<T> = {
+  fn: () => Promise<T>;
+  exitChain: boolean;
+  context?: RequestContext;
 };
+
+type Middleware<T> = <T>(options: MiddlewareInOut) => MiddlewareInOut;
 ```
 
-The explicit way to define the default situation with 10sec timeout, circuit breaker for BTP and retry switched off would be:
+In our code we will pass around a middleware array and wrap it around evey http call at the very end:
 
 ```ts
-defaultResilienceOptions: ResilienceMiddleWareOptions = {
-  timeout: (context: RequestContext) => 10000,
-  retry: (context: RequestContext) => false,
-  circuitBreaker: enableCircuitBreakerForBTP
-};
+async function someHttpSdkMethod(middleware: Middleware[]): Promise<any> {
+  //return someHttpCall(someArgs)
+  const context = {}; //context for this method
 
-function enableCircuitBreakerForBTP(
-  context: RequestContext
-): CircuitBreakerOptions {
-  if (context.url.includes(btpDomain)) {
-    return defaultCircuitBreakerOptions;
-  }
-  return false;
+  return joinMiddleware(middleware)({
+    context,
+    fn: someHttpCall,
+    someArgs,
```

Comment:
I initially did it this way, but I guess the circuit breaker needs the arguments to work better (not sure) and it could be valuable for the context as well. But valid point for discussion.